File: src/install/PackageManager/PopulateManifestCache.zig
const StartManifestTaskError = bun.OOM || error{InvalidURL};
fn startManifestTask(manager: *PackageManager, pkg_name: []const u8, dep: *const Dependency, needs_extended_manifest: bool) StartManifestTaskError!void {
    const task_id = Task.Id.forManifest(pkg_name);
    if (manager.hasCreatedNetworkTask(task_id, dep.behavior.optional)) {
        return;
    }
    manager.startProgressBarIfNone();
    var task = manager.getNetworkTask();
    task.* = .{
        .package_manager = manager,
        .callback = undefined,
        .task_id = task_id,
        .allocator = manager.allocator,
    };
    try task.forManifest(pkg_name, manager.allocator, manager.scopeForPackageName(pkg_name), null, dep.behavior.optional, needs_extended_manifest);
    manager.enqueueNetworkTask(task);
}

const Packages = union(enum) {
    all,
    ids: []const PackageID,
};

/// Populate the manifest cache for packages included from `root_pkg_ids`. Only manifests of
/// direct dependencies of the `root_pkg_ids` are populated. If `root_pkg_ids` has length 0
/// all packages in the lockfile will have their manifests fetched if necessary.
pub fn populateManifestCache(manager: *PackageManager, packages: Packages) !void {
    const log_level = manager.options.log_level;
    const lockfile = manager.lockfile;
    const resolutions = lockfile.buffers.resolutions.items;
    const dependencies = lockfile.buffers.dependencies.items;
    const string_buf = lockfile.buffers.string_bytes.items;
    const pkgs = lockfile.packages.slice();
    const pkg_resolutions = pkgs.items(.resolution);
    const pkg_names = pkgs.items(.name);
    const pkg_dependencies = pkgs.items(.dependencies);

    switch (packages) {
        .all => {
            var seen_pkg_ids: std.AutoHashMap(PackageID, void) = .init(manager.allocator);
            defer seen_pkg_ids.deinit();

            for (dependencies, 0..) |*dep, _dep_id| {
                const dep_id: DependencyID = @intCast(_dep_id);

                const pkg_id = resolutions[dep_id];
                if (pkg_id == invalid_package_id) {
                    continue;
                }

                if ((try seen_pkg_ids.getOrPut(pkg_id)).found_existing) {
                    continue;
                }

                const res = pkg_resolutions[pkg_id];
                if (res.tag != .npm) {
                    continue;
                }

                const pkg_name = pkg_names[pkg_id];
                const needs_extended_manifest = manager.options.minimum_release_age_ms != null;

                _ = manager.manifests.byName(
                    manager,
                    manager.scopeForPackageName(pkg_name.slice(string_buf)),
                    pkg_name.slice(string_buf),
                    .load_from_memory_fallback_to_disk,
                    needs_extended_manifest,
                ) orelse {
                    try startManifestTask(manager, pkg_name.slice(string_buf), dep, needs_extended_manifest);
                };

                manager.flushNetworkQueue();
                _ = manager.scheduleTasks();
            }
        },
        .ids => |ids| {
            for (ids) |root_pkg_id| {
                const pkg_deps = pkg_dependencies[root_pkg_id];
                for (pkg_deps.begin()..pkg_deps.end()) |dep_id| {
                    if (dep_id >= dependencies.len) continue;
                    const pkg_id = resolutions[dep_id];
                    if (pkg_id == invalid_package_id) continue;
                    const dep = &dependencies[dep_id];

                    const resolution: Resolution = pkg_resolutions[pkg_id];
                    if (resolution.tag != .npm) continue;

                    const needs_extended_manifest = manager.options.minimum_release_age_ms != null;
                    const package_name = pkg_names[pkg_id].slice(string_buf);
                    _ = manager.manifests.byName(
                        manager,
                        manager.scopeForPackageName(package_name),
                        package_name,
                        .load_from_memory_fallback_to_disk,
                        needs_extended_manifest,
                    ) orelse {
                        try startManifestTask(manager, package_name, dep, needs_extended_manifest);

                        manager.flushNetworkQueue();
                        _ = manager.scheduleTasks();
                    };
                }
            }
        },
    }

    manager.flushNetworkQueue();
    _ = manager.scheduleTasks();

    if (manager.pendingTaskCount() > 0) {
        const RunClosure = struct {
            manager: *PackageManager,
            err: ?anyerror = null,
            pub fn isDone(closure: *@This()) bool {
                closure.manager.runTasks(
                    *PackageManager,
                    closure.manager,
                    .{
                        .onExtract = {},
                        .onResolve = {},
                        .onPackageManifestError = {},
                        .onPackageDownloadError = {},
                        .progress_bar = true,
                        .manifests_only = true,
                    },
                    true,
                    closure.manager.options.log_level,
                ) catch |err| {
                    closure.err = err;
                    return true;
                };

                return closure.manager.pendingTaskCount() == 0;
            }
        };

        var run_closure: RunClosure = .{ .manager = manager };
        manager.sleepUntil(&run_closure, &RunClosure.isDone);

        if (log_level.showProgress()) {
            manager.endProgressBar();
            Output.flush();
        }

        if (run_closure.err) |err| {
            return err;
        }
    }
}

/// After resolution, verify every npm-tagged package in the lockfile
/// satisfies the configured `minimumReleaseAge` cooldown.
///
/// The resolution-time filter (`findBestVersionWithFilter`, etc.) only
/// runs when Bun is actually picking a new version. If the lockfile
/// already pins a version — e.g. it was resolved before the cooldown
/// was configured, or by a developer whose local bunfig was less strict
/// — that install path skips the filter entirely. Without this gate,
/// `bun install` (and `bun install --frozen-lockfile`) will happily
/// install a locked version that was published inside the cooldown
/// window, defeating the supply-chain protection the setting is meant
/// to provide.
///
/// This loads manifests for every locked npm package, looks up the
/// exact pinned version's publish timestamp, and aggregates every
/// violation into `manager.log` as an error. Excludes from
/// `minimumReleaseAgeExcludes` are honored.
pub fn enforceLockfileAgeFilter(manager: *PackageManager) !void {
    const min_age_ms = manager.options.minimum_release_age_ms orelse return;

    // Make sure manifests are loaded from disk / network before we
    // inspect publish timestamps. `populateManifestCache` already
    // honors `minimum_release_age_ms` by requesting extended manifests.
    try populateManifestCache(manager, .all);

    const lockfile = manager.lockfile;
    const pkgs = lockfile.packages.slice();
    const pkg_resolutions = pkgs.items(.resolution);
    const pkg_names = pkgs.items(.name);
    const pkg_name_hashes = pkgs.items(.name_hash);
    const string_buf = lockfile.buffers.string_bytes.items;
    const min_age_seconds = min_age_ms / std.time.ms_per_s;

    for (pkg_resolutions, pkg_names, pkg_name_hashes) |resolution, name, name_hash| {
        if (resolution.tag != .npm) continue;

        const name_str = name.slice(string_buf);

        // Fail closed: if we cannot reach the manifest or locate the exact
        // pinned version, we cannot prove the version satisfies the cooldown.
        // Silently skipping would re-open the lockfile bypass this gate is
        // meant to close (e.g. a version that was unpublished from the
        // registry, or a manifest fetch that couldn't be completed).
        const manifest = manager.manifests.byNameHash(
            manager,
            manager.scopeForPackageName(name_str),
            name_hash,
            .load_from_memory_fallback_to_disk,
            true,
        ) orelse {
            if (isExcludedByName(name_str, manager.options.minimum_release_age_excludes)) continue;
            manager.log.addErrorFmt(
                null,
                logger.Loc.Empty,
                manager.allocator,
                "Package \"{s}@{f}\" in lockfile could not be checked against minimum release age (manifest unavailable)",
                .{ name_str, resolution.value.npm.version.fmt(string_buf) },
            ) catch bun.outOfMemory();
            continue;
        };

        if (manifest.shouldExcludeFromAgeFilter(manager.options.minimum_release_age_excludes)) continue;

        const find_result = manifest.findByVersion(resolution.value.npm.version) orelse {
            manager.log.addErrorFmt(
                null,
                logger.Loc.Empty,
                manager.allocator,
                "Package \"{s}@{f}\" in lockfile could not be checked against minimum release age (version not in manifest)",
                .{ name_str, resolution.value.npm.version.fmt(string_buf) },
            ) catch bun.outOfMemory();
            continue;
        };
        if (!Npm.PackageManifest.isPackageVersionTooRecent(find_result.package, min_age_ms)) continue;

        manager.log.addErrorFmt(
            null,
            logger.Loc.Empty,
            manager.allocator,
            "Package \"{s}@{f}\" in lockfile was published within minimum release age of {d} seconds",
            .{
                name_str,
                resolution.value.npm.version.fmt(string_buf),
                min_age_seconds,
            },
        ) catch bun.outOfMemory();
    }
}

/// Mirrors `PackageManifest.shouldExcludeFromAgeFilter` for the code path
/// where no manifest is available (the manifest lookup above returned null).
/// Kept in sync with the real check in `src/install/npm.zig`.
fn isExcludedByName(name: []const u8, exclusions: ?[]const []const u8) bool {
    const excl = exclusions orelse return false;
    for (excl) |entry| {
        if (bun.strings.eql(entry, name)) return true;
    }
    return false;
}

const std = @import("std");

const bun = @import("bun");
const Output = bun.Output;
const logger = bun.logger;

const Dependency = bun.install.Dependency;
const DependencyID = bun.install.DependencyID;
const Npm = bun.install.Npm;
const PackageID = bun.install.PackageID;
const PackageManager = bun.install.PackageManager;
const Resolution = bun.install.Resolution;
const Task = bun.install.Task;
const invalid_package_id = bun.install.invalid_package_id;

File: src/install/PackageManager/PopulateManifestCache.rs
use crate::lockfile::package::PackageColumns as _;
use bun_collections::HashMap;
use bun_core::Output;

use crate::NetworkTask;
use crate::PackageID;
use crate::invalid_package_id;
use crate::lockfile::Lockfile;
// Import the
// *module* under the `Task` name so `Task::Id` resolves as a path (matches
// `run_tasks.rs` / `PackageManagerEnqueue.rs`).
use super::PackageManager;
use super::enqueue;
use super::run_tasks;
use crate::package_manager_task as Task;
use crate::resolution::Tag as ResolutionTag;

#[derive(thiserror::Error, strum::IntoStaticStr, Debug)]
pub(crate) enum StartManifestTaskError {
    #[error("OutOfMemory")]
    OutOfMemory,
    #[error("InvalidURL")]
    InvalidURL,
}
bun_core::oom_from_alloc!(StartManifestTaskError);
impl From<crate::network_task::ForManifestError> for StartManifestTaskError {
    fn from(e: crate::network_task::ForManifestError) -> Self {
        match e {
            crate::network_task::ForManifestError::OutOfMemory => Self::OutOfMemory,
            crate::network_task::ForManifestError::InvalidURL => Self::InvalidURL,
        }
    }
}
impl From<StartManifestTaskError> for crate::Error {
    fn from(e: StartManifestTaskError) -> Self {
        match e {
            StartManifestTaskError::OutOfMemory => crate::Error::Alloc(bun_alloc::AllocError),
            StartManifestTaskError::InvalidURL => crate::Error::InvalidURL,
        }
    }
}

/// `is_required`: a failed fetch is logged as an error rather than a warning.
fn start_manifest_task(
    manager: &mut PackageManager,
    pkg_name: &[u8],
    is_required: bool,
    needs_extended_manifest: bool,
) -> Result<(), StartManifestTaskError> {
    // best-effort metadata backfill: nothing to do without the network
    if manager.options.offline == crate::package_manager_real::options::OfflineMode::Offline {
        return Ok(());
    }
    let task_id = Task::Id::for_manifest(pkg_name);
    if run_tasks::has_created_network_task(manager, task_id, is_required) {
        return Ok(());
    }
    if manager.options.log_level.show_progress() {
        manager.start_progress_bar_if_none();
    }

    let mut task = NetworkTask::new(task_id, manager);
    {
        let PackageManager {
            log, env, options, ..
        } = &mut *manager;
        let scope = options.scope_for_package_name(pkg_name);
        task.for_manifest(
            log,
            env.get(),
            pkg_name,
            scope,
            None,
            !is_required,
            needs_extended_manifest,
        )?;
    }

    enqueue::enqueue_network_task(manager, task);
    Ok(())
}

#[derive(Clone, Copy)]
pub enum Packages<'a> {
    /// Every npm package in this lockfile (the one a migration is building,
    /// not yet `manager.lockfile`); best-effort backfill, so failures are warnings.
    All(&'a Lockfile),
    /// The direct dependencies of these workspace packages; a required one failing is an error, see [`print_fetch_failures`].
    Ids(&'a [PackageID]),
    /// The manifests of these packages themselves (by name), not of their dependencies; best-effort, so failures are warnings.
    Exact(&'a [PackageID]),
}

/// `RunTasksCtx` for the hook-less `run_tasks` call in
/// `populate_manifest_cache`.
struct ManifestsOnlyCtx<'a>(&'a mut PackageManager);
impl run_tasks::RunTasksCtx for ManifestsOnlyCtx<'_> {
    fn manager(&mut self) -> &mut PackageManager {
        self.0
    }
    fn progress_bar(&self) -> bool {
        true
    }
    fn manifests_only(&self) -> bool {
        true
    }
}

/// An npm package whose manifest may need fetching: its name copied out of the
/// lockfile string buffer so the manager can be mutated while it is in use.
struct ManifestCandidate {
    name: NameBuf,
    is_required: bool,
}

/// npm package names are at most 214 bytes; anything longer cannot have come
/// from a registry and is fetched with a heap copy instead.
#[allow(clippy::large_enum_variant)] // the inline buffer is the point
enum NameBuf {
    Inline { buf: [u8; 214], len: u8 },
    Heap(Box<[u8]>),
}

impl NameBuf {
    fn new(name: &[u8]) -> NameBuf {
        if name.len() <= 214 {
            let mut buf = [0u8; 214];
            buf[..name.len()].copy_from_slice(name);
            NameBuf::Inline {
                buf,
                len: name.len() as u8,
            }
        } else {
            NameBuf::Heap(Box::from(name))
        }
    }
    fn as_slice(&self) -> &[u8] {
        match self {
            NameBuf::Inline { buf, len } => &buf[..*len as usize],
            NameBuf::Heap(b) => b,
        }
    }
}

fn npm_candidate(
    lockfile: &Lockfile,
    pkg_id: PackageID,
    is_required: bool,
) -> Option<ManifestCandidate> {
    if lockfile.packages.items_resolution()[pkg_id as usize].tag != ResolutionTag::Npm {
        return None;
    }
    let name = lockfile.packages.items_name()[pkg_id as usize]
        .slice(lockfile.buffers.string_bytes.as_slice());
    Some(ManifestCandidate {
        name: NameBuf::new(name),
        is_required,
    })
}

fn fetch_manifest_if_uncached(
    manager: &mut PackageManager,
    candidate: &ManifestCandidate,
) -> crate::Result<()> {
    let cache_ctx = manager.manifest_disk_cache_ctx();
    let name = candidate.name.as_slice();
    let needs_extended_manifest = manager.options.minimum_release_age_ms.is_some();
    let scope = manager.options.scope_for_package_name(name);
    let cached = manager
        .manifests
        .by_name(cache_ctx, scope, name, needs_extended_manifest);
    if cached.is_none() {
        start_manifest_task(
            manager,
            name,
            candidate.is_required,
            needs_extended_manifest,
        )?;
        run_tasks::flush_network_queue(manager);
        let _ = run_tasks::schedule_tasks(manager);
    }
    Ok(())
}

/// Populate the manifest cache for the packages `packages` selects (see
/// [`Packages`]).
pub fn populate_manifest_cache(
    manager: &mut PackageManager,
    packages: Packages<'_>,
) -> crate::Result<()> {
    let log_level = manager.options.log_level;

    match packages {
        Packages::All(lockfile) => {
            let mut seen_pkg_ids: HashMap<PackageID, ()> = HashMap::new();

            for dep_id in 0..lockfile.buffers.dependencies.len() {
                let pkg_id = lockfile.buffers.resolutions[dep_id];
                if pkg_id == invalid_package_id {
                    continue;
                }

                if seen_pkg_ids.insert(pkg_id, ()).is_some() {
                    continue;
                }

                let Some(candidate) = npm_candidate(lockfile, pkg_id, false) else {
                    continue;
                };
                fetch_manifest_if_uncached(manager, &candidate)?;
                // `.All` flushes after every candidate, cached or not.
                run_tasks::flush_network_queue(manager);
                let _ = run_tasks::schedule_tasks(manager);
            }
        }
        Packages::Ids(ids) => {
            for &root_pkg_id in ids {
                let pkg_deps = manager.lockfile.packages.items_dependencies()[root_pkg_id as usize];
                for dep_id in pkg_deps.begin()..pkg_deps.end() {
                    let dep_id = dep_id as usize;
                    let (pkg_id, is_required) = {
                        let buffers = &manager.lockfile.buffers;
                        if dep_id >= buffers.dependencies.len() {
                            continue;
                        }
                        let pkg_id = buffers.resolutions[dep_id];
                        if pkg_id == invalid_package_id {
                            continue;
                        }
                        (pkg_id, buffers.dependencies[dep_id].behavior.is_required())
                    };
                    let Some(candidate) = npm_candidate(&manager.lockfile, pkg_id, is_required)
                    else {
                        continue;
                    };
                    fetch_manifest_if_uncached(manager, &candidate)?;
                }
            }
        }
        Packages::Exact(ids) => {
            for &pkg_id in ids {
                let Some(candidate) = npm_candidate(&manager.lockfile, pkg_id, false) else {
                    continue;
                };
                fetch_manifest_if_uncached(manager, &candidate)?;
            }
        }
    }

    run_tasks::flush_network_queue(manager);
    let _ = run_tasks::schedule_tasks(manager);

    if run_tasks::pending_task_count(manager) > 0 {
        let mut err: Option<crate::Error> = None;
        PackageManager::sleep_until(manager, |manager| {
            let log_level = manager.options.log_level;
            if let Err(e) = run_tasks::run_tasks(&mut ManifestsOnlyCtx(manager), true, log_level) {
                err = Some(e);
                return true;
            }
            run_tasks::pending_task_count(manager) == 0
        });

        if log_level.show_progress() {
            manager.end_progress_bar();
            Output::flush();
        }

        if let Some(err) = err {
            return Err(err);
        }
    }

    Ok(())
}

/// Prints the fetch failures a [`Packages::Ids`] pass logged; true when one of them is a required dependency's.
pub fn print_fetch_failures(manager: &mut PackageManager) -> crate::Result<bool> {
    let log = &mut manager.log;
    let failed_required = log.has_errors();
    if !log.msgs.is_empty() {
        Output::flush();
        log.print(core::ptr::from_mut(Output::error_writer()))?;
        log.reset();
    }
    Ok(failed_required)
}
